### ISSUE TITLE
Help need

### ISSUE THREAD

**Issue by renchili** [VictoriaMetrics/VictoriaMetrics#5993]
### Is your question request related to a specific component?

vmcluster , vmstorage bad performance under certain senarios.

### Describe the question in detail

As we store 16months data in a single vmstorage cluster, and take bill from the cloud vendor, we are hard to estimate the disk size for the coming months. So we will expand the  storage disk time, to time. 

When we operating the cluster, it will have this:
high iowait **after** the expansion
![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/33677006/6d41847a-9160-4121-a04f-6db0139b5497)

We looked in more details :
![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/33677006/30b19410-4d0b-47fd-9fb0-2ef92595fd6c)
it will start some sort of indexdb/file merging behaviours.
We have around 1.6T indexdb/file 

We compared with another cluster( only have few months data), 
![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/33677006/da290098-d53d-4caa-b33b-1190cf78cbf5)
then it wont have much impact after our operation.

Beside  as the disk capacity will decreasing dramatically during the operation, we have build some sort of HA.
We will switch the  select query to a back cluster, then switch it back after operation.
But as we switch back,  same issue got worse：
![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/33677006/db692f43-d77a-40bc-8a32-7419e0e57264)
Some unknow points caused huge writing capacity downgrade.

We are seeking help to understand what happened.

For your information:
vmagent: 
 - version:1.91.3
 - number: 22, in two group, 9 + 13
 - server type: n2-highcpu-64
 - cpu: 64, memory: 64
 - os: CentOS Linux release 7.9.2009 (Core) 5.12.12-1.el7.elrepo.x86_64
 - replicate = 2
vminsert:
  - version: 1.91.3-cluster
  - number: 10
  - server type: n2-highcpu-32
  - cpu: 32, memory: 32
  - os: CentOS Linux release 7.9.2009 (Core) 5.12.12-1.el7.elrepo.x86_64)
  - replicate = 2  
vmstorage:
  - version: 1.91.3-cluster
  - server type: n2-highmem-16
  - cpu:16, memory: 128
  - os: CentOS Linux release 7.9.2009 (Core) 5.12.12-1.el7.elrepo.x86_64)
vmselect:
  - version: 1.91.3-cluster
  - number: 11 in the same az, and 11 in another az  (Will this cause any issue? We been warned by the cloud vendor cross az traffice quality issue)
  - server type: n2-highmem-16
  - cpu: 16, memory: 128 
  
  Or data scale is : Active timeseries 7.47Million, total datapoints: 123 trillion， ingestion Rate：1.57 mil/s
  , datapoints average size: 0.74B

### Troubleshooting docs

- [ ] General - https://docs.victoriametrics.com/Troubleshooting.html
- [ ] vmagent - https://docs.victoriametrics.com/vmagent.html#troubleshooting
- [ ] vmalert - https://docs.victoriametrics.com/vmalert.html#troubleshooting

**Comment by mmkkoo** [comment]
@valyala I also have this problem，please take a look

**Comment by hagen1778** [comment]
Hello @renchili!
> high iowait after the expansion

According to [documentation](https://docs.victoriametrics.com/#storage), vmstorage performs background merges for data and index parts. The merges improve compression and reduce number of files on disk. To perform a merge, vmstorage locates at least 2 data parts of comparable size and merges them into a new part. Once merge is complete, the old 2 data parts are deleted. This means that to perform a merge vmstorage will have to occupy some additional disk space until the merge is done.

Before doing a merge, vmstorage estimates the amount of free disk space. If it is not enough of disk space for merging, the merge will be postponed:
> VictoriaMetrics doesn’t merge parts if their summary size exceeds free disk space. This prevents from potential out of disk space errors during merge. The number of parts may significantly increase over time under free disk space shortage. This increases overhead during data querying, since VictoriaMetrics needs to read data from bigger number of parts per each request. That’s why it is recommended to have at least 20% of free disk space under directory pointed by -storageDataPath command-line flag.

It is very likely that in your case a lot of merges were postponed due to lack of disk space. Once you extend disk space those merges start to happen. Merge is IO intensive operation - it requires read and write throughput. It could saturate disk IOPS limits, especially for slow disks. This is why you observe high iowait after expanding the disk.

In 1.91.3 you can reduce the intensity of merges by limiting the number of threads doing the merge via `smallMergeConcurrency` cmd-line flag. Please note, the merging logic has changed since then and this flag was deprecated in newer versions.

I'd suggest to perform following actions:
1. Always keep 20% of free disk space for merges to happen in time
2. Start using disks with higher IO throughput for merges to not saturate IO completely
3. Decrease `smallMergeConcurrency` if you plan to stay at 1.91.X versions
4. Update to newer version of VM

**Comment by renchili** [comment]
> Hello @renchili!
> 
> > high iowait after the expansion
> 
> 
> 
> According to [documentation](https://docs.victoriametrics.com/#storage), vmstorage performs background merges for data and index parts. The merges improve compression and reduce number of files on disk. To perform a merge, vmstorage locates at least 2 data parts of comparable size and merges them into a new part. Once merge is complete, the old 2 data parts are deleted. This means that to perform a merge vmstorage will have to occupy some additional disk space until the merge is done.
> 
> 
> 
> Before doing a merge, vmstorage estimates the amount of free disk space. If it is not enough of disk space for merging, the merge will be postponed:
> 
> > VictoriaMetrics doesn’t merge parts if their summary size exceeds free disk space. This prevents from potential out of disk space errors during merge. The number of parts may significantly increase over time under free disk space shortage. This increases overhead during data querying, since VictoriaMetrics needs to read data from bigger number of parts per each request. That’s why it is recommended to have at least 20% of free disk space under directory pointed by -storageDataPath command-line flag.
> 
> 
> 
> It is very likely that in your case a lot of merges were postponed due to lack of disk space. Once you extend disk space those merges start to happen. Merge is IO intensive operation - it requires read and write throughput. It could saturate disk IOPS limits, especially for slow disks. This is why you observe high iowait after expanding the disk.
> 
> 
> 
> In 1.91.3 you can reduce the intensity of merges by limiting the number of threads doing the merge via `smallMergeConcurrency` cmd-line flag. Please note, the merging logic has changed since then and this flag was deprecated in newer versions.
> 
> 
> 
> I'd suggest to perform following actions:
> 
> 1. Always keep 20% of free disk space for merges to happen in time
> 
> 2. Start using disks with higher IO throughput for merges to not saturate IO completely
> 
> 3. Decrease `smallMergeConcurrency` if you plan to stay at 1.91.X versions
> 
> 4. Update to newer version of VM
> 
> 

hi thanks for your help，

But how we know there are pending merges that is not finished and how to escape at first moment of upgrade？ such a smoth switching to higher version？

**Comment by hagen1778** [comment]
> But how we know there are pending merges that is not finished？

If you have less than 200GiB of free space it is very likely there are some merges that were postponed. It is ok for some merges to be postponed, it is not ok if there are too many merges postponed. There were a metric that showed the number of postponed merges `vm_merge_need_free_disk_space`, but it was deprecated with time - see https://github.com/VictoriaMetrics/VictoriaMetrics/issues/686

In general, you should try to keep 20% of disk space free. This is the optimal size for merges and for disk performance. 

> how to escape at first moment of upgrade?

There is no way to delay the merges. You can limit the speed of merges as I mentioned above. Or increase disk IO limits so it don't get saturated. Or switch to newer versions of VM which has altered merging logic.

> such a smoth switching to higher version？

I didn't get this question.

**Comment by renchili** [comment]
> > But how we know there are pending merges that is not finished？
> 
> 
> 
> If you have less than 200GiB of free space it is very likely there are some merges that were postponed. It is ok for some merges to be postponed, it is not ok if there are too many merges postponed. There were a metric that showed the number of postponed merges `vm_merge_need_free_disk_space`, but it was deprecated with time - see https://github.com/VictoriaMetrics/VictoriaMetrics/issues/686
> 
> 
> 
> In general, you should try to keep 20% of disk space free. This is the optimal size for merges and for disk performance. 
> 
> 
> 
> > how to escape at first moment of upgrade?
> 
> 
> 
> There is no way to delay the merges. You can limit the speed of merges as I mentioned above. Or increase disk IO limits so it don't get saturated. Or switch to newer versions of VM which has altered merging logic.
> 
> 
> 
> > such a smoth switching to higher version？
> 
> 
> 
> I didn't get this question.
> 
> 
> 
> 

hmm I'm afraid we miscommunicated.

First the io saturation is happened when we start new version not on **old** version.

And second for more information
we never let the vmstore lace than 1TB free disck space well this some time is less than 4% free space to make 20% is nearly 5TB
our data is roughly increasing 1-1.5TB per month and rotation period is 16months （the cluster not yet met the disk balance， in out size balance）
